Bug 1833195: delete bundle objects after CSV gets deleted

exdx opened the pull request:
<!--

Before making a PR, please read our contributing guidelines https://github.com/operator-framework/operator-lifecycle-manager/blob/master/CONTRIBUTING.MD

Note: Make sure your branch is rebased to the latest upstream master.

-->

**Description of the change:**
OLM now supports secrets and configmaps in the bundle. These should have ownerrefs associated with the CSV in the bundle. When the CSV is deleted, these objects should be deleted as well. Since they are namespaced scoped, the kube gc will delete them automatically. 

**Motivation for the change:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive


<!--

Note: If this PR is fixing an issue make sure to add a note saying:
Closes #<ISSUE_NUMBER>

-->

exdx commented:
/retest

awgreene commented:
/retest

awgreene commented on test/e2e/gc_e2e_test.go:
```diff
@@ -3,24 +3,22 @@ package e2e
 	"github.com/blang/semver"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/ownerutil"
+	"github.com/operator-framework/operator-lifecycle-manager/test/e2e/ctx"
+	. "github.com/operator-framework/operator-lifecycle-manager/test/e2e/dsl"
```
OLM resources are typically imported in a third import block

exdx replied:
This was the result of `go fmt` but I can adjust it

awgreene commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +282,91 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+			Expect(err).NotTo(HaveOccurred())
+
+			// wait for csv install to complete
+			Eventually(func() bool {
+				csv, _ := operatorClient.OperatorsV1alpha1().ClusterServiceVersions(testNamespace).Get(context.TODO(), fetchedOwner.GetName(), metav1.GetOptions{})
+				return csv.Status.Phase == v1alpha1.CSVPhaseSucceeded
+			}).Should(BeTrue())
```
Alternatively, there is a [fetchCSV](https://github.com/operator-framework/operator-lifecycle-manager/blob/d23b7ab12e3925a16bdb961cc9f8b8ddbad01cce/test/e2e/csv_e2e_test.go#L3409) function which waits for a CSV to reach the desired phase.

awgreene commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +282,91 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+				Data: map[string][]byte{"name": []byte("Alice")},
+			}
+
+			// set ownerrefs to objects
+			ownerutil.AddOwner(cfg, fetchedOwner, true, false)
+			ownerutil.AddOwner(s, fetchedOwner, true, false)
```
Nit: This tests ensures that secrets that are owned by the CSV are GC'd when the CSV is deleted, but it does not ensure that the Secrets have the ownerrefs. Could come back to bite us.

exdx replied:
That is a good point, I'm wondering if there is a need for an additional follow-up test, I'll look into our existing tests and see if something like this exists. I imagine it would be a csv e2e test.

awgreene commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +282,91 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+		})
+
+		When("CSV is deleted", func() {
+
```
Nit: Please remove the newline.

awgreene commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +282,91 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+			})
+
+			It("should delete the associated configmap and secret", func() {
+				_, err := kubeClient.GetConfigMap(testNamespace, configMap.GetName())
+				Expect(err).To(HaveOccurred())
+				Expect(k8serrors.IsNotFound(err)).To(BeTrue())
+				ctx.Ctx().Logf("configmap successfully garbage collected after csv deleted")
+
+				_, err = kubeClient.GetSecret(testNamespace, secret.GetName())
+				Expect(err).To(HaveOccurred())
+				Expect(k8serrors.IsNotFound(err)).To(BeTrue())
+				ctx.Ctx().Logf("secret successfully garbage collected after csv deleted")
```
We may want to put this in an `Eventually` block.

exdx replied:
Originally I did have these in `Eventually` blocks, but I realized checking it this way is actually more precise. Since we are using the foreground deletion option when deleting the CSV, that ensures that dependent objects (the secret and configmap) are GC'd before the CSV itself. Therefore once the CSV is gone we should be 100% sure that the dependent objects are gone as well. Checking it this way confirms that to be the case

awgreene replied:
Ah great point!

awgreene commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +285,104 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+			source, err = operatorClient.OperatorsV1alpha1().CatalogSources(source.GetNamespace()).Create(context.TODO(), source, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred(), "could not create catalog source")
+
+			// wait for catalog pods to come up
+			selector := labels.SelectorFromSet(map[string]string{"olm.catalogSource": sourceName})
+			_, err = awaitPods(GinkgoT(), kubeClient, source.GetNamespace(), selector.String(), podCount(1))
+			Expect(err).ToNot(HaveOccurred(), "catalog source pod not found")
```
Do we need to wait for the pod to come up? Won't the subscription eventually succeed?

exdx replied:
That's an interesting thought - we could create them in parallel. I'm a little concerned about that though because I believe `Eventually` has a default 60 second timeout - its possible on a slower cluster it could flake on CI. Let me see if I can set the timeout.

exdx replied:
Ok so I addressed this by relying on the default subscription checker polling which is 5 minutes. It did make the code cleaner this way, so nice catch.

awgreene commented:
Unit tests are failing.

exdx commented:
> Unit tests are failing.

didn't check things again after rebasing off master 🤦

benluddy commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +284,101 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+			}).Should(BeNil())
+
+			// confirm extra bundle objects (secret and configmap) are installed
+			Eventually(func() bool {
+				_, err := kubeClient.GetSecret(testNamespace, secretName)
+				return k8serrors.IsNotFound(err)
+			}).Should(BeFalse())
```
This assertion would pass if err was some non-nil value other than a not found error. Also, the message in the test report on failures would be something like "expected False but was True". Something like this:

```suggestion
			Eventually(func() error {
				_, err := kubeClient.GetSecret(testNamespace, secretName)
				return err
			}).Should(Succeed())
```
should give a more helpful failure message like "expected no error but was 'some error'".

harishsurf commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +284,101 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+				return err
+			}).Should(BeNil())
+
+			// confirm extra bundle objects (secret and configmap) are installed
```
Since you are verifying that your `CatalogSource` and `Subscription`  which is part of your _pre-setup_ created `secret` & `configMap`, it could be part of its own `It` spec.  For example, your test could look something like: 
```
When("a catalog source is created using grpc and a subscription", func() {
         BeforeEach(func(){
         // create catsrc and subscription  
         })
         It("should have created a secret and a configmap", func() {
                 // validate these objects
         })
```

harishsurf commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +284,101 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+			packageName   = "busybox"
+			channelName   = "alpha"
+			subName       = "test-subscription"
+			csvName       = "busybox.v2.0.0"
```
It is better to declare variables in the Container blocks where they are used. For example, `csvName` could be declared inside of `When` block at line 346.

harishsurf commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +284,101 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+
+		var (
+			source *v1alpha1.CatalogSource
+			err    error
```
We only want to declare variables at Container level only if we want to share them between multiple `It` specs. err isn't something we would want to share across multiple specs. So it is advisable to declare it right before it is used. [See here](https://github.com/operator-framework/operator-lifecycle-manager/blob/db290a7d8503dcfa8f49cb976bcc3da1b054392b/test/e2e/metrics_e2e_test.go#L60)

exdx commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +284,101 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+
+		var (
+			source *v1alpha1.CatalogSource
+			err    error
```
Agreed - I hate declaring the error at the top scope (very unidiomatic) but I think I got bit by having it at a local scope in another test and changed it. Will change it back. Still getting the hang of the framework

harishsurf commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +284,101 @@ var _ = Describe("Garbage collection for dependent resources", func() {
 
 	})
 
+	Context("When installing a bundle with configmap and secret objects these objects should be deleted when the CSV is deleted", func() {
```
the `should be deleted when CSV is deleted` could just be part of the `It` spec description. 
The way the Container block descriptions are written is :
`Describe` blocks to describe the individual behaviors of your code and `Context` blocks exercise those behaviors under different circumstances. You could additionally use `When` for better readability. 
For example:
```
[Describe] Subscriptions create required objects from Catalogs
  [Context] Given a Namespace with an OwnNamespace OperatorGroup
      [When] a CatalogSource is created with a bundle that contains prometheus objects
         [It] should install the operator successfully
             [By] creating a subscription to the CatalogSource
         [It] should have created the expected prometheus objects
```
Also, note that: `Describe/ Context / When` are functionally identical and only differ in their names.

exdx commented:
Added two additional e2e tests for the 2 upgrade scenarios @ecordell mentioned.

Also added the ownership code to service objects in the bundle as well for consistency across all bundle objects OLM resolves.

exdx commented:
/retest

ecordell commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +284,280 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+
+			It("OLM should have put new configmap in place", func() {
+				Eventually(func() error {
+					_, err := kubeClient.GetConfigMap(testNamespace, upgradedConfigMapName)
```
should we check that the old one was deleted?

exdx replied:
we can check that, sure

exdx commented:
Tests should pass once Quay is back in a healthy state

exdx commented:
/retest

ecordell commented:
/lgtm

benluddy commented on test/e2e/gc_e2e_test.go:
```diff
@@ -284,4 +284,285 @@ var _ = Describe("Garbage collection for dependent resources", func() {
+			})
+
+			It("OLM should have upgraded associated configmap in place", func() {
+				Eventually(func() bool {
+					cfg, err := kubeClient.GetConfigMap(testNamespace, configmapName)
+					if err != nil {
+						return false
+					}
+					// check data in configmap to ensure it is the new data (configmap was updated in the newer bundle)
+					// new value in the configmap is "updated-very-much"
+					data := cfg.Data["special.how"]
+					return data == "updated-very-much"
+				}).Should(BeTrue())
```
I think both Expect and Eventually/Consistently can handle multiple return values and implicitly fail if anything after the first return value is non-nil, which would let you write this:

```suggestion
				Eventually(func() (string, error) {
					cfg, err := kubeClient.GetConfigMap(testNamespace, configmapName)
					if err != nil {
						return "", err
					}
					return cfg.Data["special.how"], nil
				}).Should(Equal("updated-very-much"))
```

exdx replied:
nice
